### PR TITLE
feat: Implement design body

### PR DESCRIPTION
**Pull request by justinkambic**
## Summary

Related to #87.

## Implementation details

Adds an updated body layout per our latest design.

image

## How to validate this change

This is a largely visual change, but there will be many revisions to the functionality throughout the recorder. I'll include detailed instructions when I open this for review.

### CODE REVIEW

**Review by kyungeunni — COMMENTED** [review]
Looking great ✨ 

Just a minor style issue when viewport width is smaller than 850px
image

File: src/components/TestResult/ResultErrorBody.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+*/
+
+import { EuiFlexGroup, EuiFlexItem, EuiText, EuiCodeBlock } from "@elastic/eui";
+import React from "react";
+import { ResultCategory } from "../../common/types";
+import { ResultErrorAccordion, symbols } from "./styles";
+
+function removeColorCodes(str = "") {
+  // eslint-disable-next-line no-control-regex
+  return str.replace(/\u001b\[.*?m/g, "");
```
Comment (lucasfcosta) [inline src/components/TestResult/ResultErrorBody.tsx:32]: How confident are we that this RegEx always catches color codes? Can't we disable the color codes when we invoke the process which yields the error message instead? That feels like it would be more reliable and less work.

I've heard a funny quote in a talk once:

> I had a problem. I tried to use a RegEx. I then had two problems.

😆

File: src/helpers/test/defaults.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+import { RecordingStatus } from "../../common/types";
+import { IRecordingContext } from "../../contexts/RecordingContext";
+import { IStepsContext } from "../../contexts/StepsContext";
+import { IUrlContext } from "../../contexts/UrlContext";
+
+export const RECORDING_CONTEXT_DEFAULTS: IRecordingContext = {
```
Comment (lucasfcosta) [inline src/helpers/test/defaults.ts:30]: Please see my comment on the other PR about new default objects for every render.

File: src/hooks/useStepsContext.test.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+          action: {
+            name: "second-step-2",
+            signals: [],
+          },
+          frameUrl: "https://www.elastic.co",
+          isMainFrame: true,
+          pageAlias: "pageAlias",
+        },
+      ],
+    ];
```
Comment (lucasfcosta) [inline src/hooks/useStepsContext.test.ts:70]: Just a nit: can we wrap the creation of steps into a function so we don't have to remember the object structure every time? There seems to be a lot of repetition here so I imagine that the following would be nice:

```
const createStep = (overrides, index) => {
  return {
    action: {
      name: `step-${index}`,
      signals: [],
    },
    frameUrl: "https://www.elastic.co",
    isMainFrame: true,
    pageAlias: "pageAlias",
    ...verrides
  };
};

const createSteps = (overridesArr) => overridesArr.map(createStep);
```

**Review by lucasfcosta — CHANGES_REQUESTED** [review]
Hey Justin, thanks a lot for the excellent work on this (as usual)! 🙏 

I have now gone through all files and most of my suggestions apply a single file or to more resilient ways of doing things.

Overall, I really like the way you've separated components, as they're now much more granular and easier to understand.

To be honest it wasn't that straightforward to follow all type references to check if we could make them a bit drier, but considering the use cases, all seem fine.

I've also pulled your branch and E2E tested it manually. I quite like that we don't have a place to type the initial URL anymore. Just clicking record and navigating away is much more intuitive and saves us weird edge cases as we had before.

Some feedback I have from that test:
* Once you click to replay steps, IMO we should immediately disable the button because it takes a little bit of time for the browser to open, and that could cause the user to click there multiple times.
* I could consistently reproduce a case in which my last step was not reproduced. Please see the example below where I had one last click on a piece of text that was never reported back.
  Screenshot 2022-02-11 at 17 51 50
* Can we hide the trailing lines after the last step on each block? Please see the trailing lines after the dots on the left in the screenshots below.
    Screenshot 2022-02-11 at 17 30 21
  Screenshot 2022-02-11 at 17 30 23



Perhaps some more E2E tests would be nice, but that's actually part of https://github.com/elastic/synthetics-recorder/issues/113 so I think we should keep that as a separate PR to avoid this one growing more, as it will get harder to review if we add too much to it.

File: e2e/tests/navigation.test.ts [hunk 4]
```diff
@@ -40,12 +40,9 @@ describe("Navigation", () => {
     await electronService.waitForPageToBeIdle();
     await electronService.navigateRecordingBrowser("http://example.com");
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply e2e/tests/navigation.test.ts:41]: This will become canon once we merge https://github.com/elastic/synthetics-recorder/pull/152.

File: src/App.css [hunk 5]
```diff
@@ -1,7 +1,8 @@
-html {
-  height: 100%;
+body {
+  height: inherit;
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/App.css:2]: Let me see what happens when I delete this from the file.

[the remark this answers is not in the document]
Comment (justinkambic) [reply src/App.css:2]: Update: I do think this is necessary, but I'm unsure why exactly. It's related to the other selector in this file. Here is how the page looks with this `body inherit` block:
image

Here's what happens when I remove that selector:

image

Note the `#root` element doesn't seem to stretch to the bottom of the page now and we see the gray background.

File: src/components/ActionDetail/FormControl.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  name: string;
+  noPrepend?: boolean;
+  onChange: ChangeEventHandler<HTMLInputElement>;
+  value: string;
+}) {
+  return (
+    <EuiFormRow label={label ?? "Command"}>
+      <EuiFieldText
+        onChange={onChange}
+        prepend={noPrepend === true ? undefined : name}
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionDetail/FormControl.tsx:47]: We're in 100% agreement about the nuanced implications of `undefined` vs `null`. Unfortunately `prepend` must be a react string or `undefined` and triggers a TS error instead. We'd need to merge an upstream update to EUI to change this. I think I probably tried to make it `null` when I wrote this code even and decided to simply use `undefined` rather than resist the API of this component.

File: src/components/ActionDetail/ActionDetail.tsx [hunk 7]
```diff
@@ -99,75 +102,79 @@ export function ActionDetail({
...
           </EuiFlexItem>
         )}
       </EuiFlexGroup>
+      <EuiSpacer />
+      <EuiFlexGroup>
+        <EuiFlexItem grow={false}>
+          <EuiButton
+            onClick={() => {
+              if (url) {
+                onURLChange(url);
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionDetail/ActionDetail.tsx:153]: I think these names are chosen to avoid overloading the actual `Setter` functions instantiated above. Like we have `setText`, `setUrl`, etc., already defined in this file, then they're referenced by these `onXChange` handlers that have some extra logic nested in them before calling the raw `Setter` function.

I'm fine to change this if it makes it easier to read, but we'd have some naming collisions if we tried doing this without renaming the existing set functions.

File: src/components/ActionElement/ActionElement.tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+          {step.action.isAssert && (
+            <Assertion
+              action={step.action}
+              actionContext={step}
+              actionIndex={actionIndex}
+              close={close}
+              onDeleteAction={onDeleteAction}
+              stepIndex={stepIndex}
+            />
+          )}
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionElement/ActionElement.tsx:118]: ++ I was unhappy with this code too, happy to make this change.

[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionElement/ActionElement.tsx]: Fixed in f4b0f68c60cac6ffabaf6e82b44f6deebcbfffc5.

File: src/components/ActionElement/ActionWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  border-left: ${props => props.theme.border.thick};
+  padding-left: 20px;
+  margin-left: 0px;
+`;
+
+interface IActionWrapper {
+  isAssert?: boolean;
+}
+
+export const ActionWrapper: React.FC<IActionWrapper> = ({
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionElement/ActionWrapper.tsx:44]: Hm - I do think this is a conversation worth having. You may have noticed that up until this PR, the entity we refer to colloquially as `Steps` was an `ActionContext[][]` throughout the codebase.

We have always referred to `ActionContext` as `Action`, even though it contains a field like `{ action: Action }`. Then we have `Assertion`, which we are basically saying is a sub-type of `ActionContext`. We should probably have a separate breakout to nail down what we want to call all of these, and do a follow-up PR where we rename the types to correspond to the formalized definitions.

As I mentioned above, in this particular case, I called this `ActionWrapper` because we think of `Action` and `Assertion` both as actions. I'm happy to rename this to `Behavior`.

File: src/components/ActionElement/AssertionHeadingText.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+export function AssertionHeadingText({ actionContext }: IAssertionHeadingText) {
+  return (
+    <EuiFlexGroup gutterSize="xs">
+      <Bold grow={false}>{actionContext.action.name}</Bold>
+      <EuiFlexItem grow={false}>
+        {
+          COMMAND_SELECTOR_OPTIONS.find(
+            option => option.value === actionContext.action.command
+          )?.text
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionElement/AssertionHeadingText.tsx:43]: Well, as things are today we will always find an item. The `command` field will always correspond to one of the `option.value` because of how they're generated. The `action.command` is a value that is chosen from `COMMAND_SELECTOR_OPTIONS` at the time of creation.

It is possible that we could have some problem in the future, so we could add a default as a sort of catch-all that should never be called. Any thoughts on what we could be in place here by default?

`Check Visibility` is an example text as shown below:

image

File: src/components/ExportScriptFlyout/Body.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+export function Body({ code, exportAsSuite, setExportAsSuite }: Props) {
+  return (
+    <EuiFlyoutBody>
+      <EuiCheckbox
+        id="export-as-suite-checkbox"
+        label="Export as suite"
+        checked={exportAsSuite}
+        onChange={() => setExportAsSuite(!exportAsSuite)}
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ExportScriptFlyout/Body.tsx:47]: Yeah it's no problem to make that change when appropriate. At this point we're still targeting the existing designs for this change, so I think we shouldn't revise it.

File: src/components/ActionStatusIndicator.tsx [hunk 12]
```diff
@@ -22,42 +22,46 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+      style={{ left: 26, top: 12, position: "relative" }}
+    >
+      <circle cx="25" cy="25" r="12" fill={euiTheme.colors.lightestShade} />
+      <circle
+        cx="25"
+        cy="25"
+        r="3"
+        fill={getColorForStatus(euiTheme, status)}
+      />
+    </svg>
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/ActionStatusIndicator.tsx:49]: :thanks

File: src/components/TestResult/ResultErrorBody.tsx (hunk 1 above)
Comment (justinkambic) [reply src/components/TestResult/ResultErrorBody.tsx:32]: Agree, not crazy about it. This code isn't new to this PR though, just moved from an old one.

Want me to simply remove this and render the raw value?

File: src/components/TestResult/TestResult.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      setStepCodeToDisplay(failedCode);
+    }
+
+    // skip procedure when there are no failed steps
+    if (steps.length && result?.failed) {
+      fetchCodeForFailure(result);
+    }
+  }, [ipc, result, setResult, steps]);
+
+  // flyout shold not show without result data
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/TestResult/TestResult.tsx]: 071bd8e94c793cad0b47acc5c42d722e5a401355

File: src/hooks/useStepsContext.test.ts (hunk 3 above)
Comment (justinkambic) [reply src/hooks/useStepsContext.test.ts:70]: Sure, I agree, I think it's awful as-is 😆

Comment (justinkambic) [reply src/hooks/useStepsContext.test.ts:70]: There's a follow-up PR behind this one that makes it much worse, adding a create function will be much nicer.

Comment (justinkambic) [reply src/hooks/useStepsContext.test.ts]: Fixed, check feda1bb4b34965db7c6dad46e6f5643e4e50dbbd.

File: src/components/StepSeparator.tsx [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+export const StepSeparatorAccordion = styled(EuiAccordion)`
+  .euiAccordion__button {
+    width: auto;
+    flex-grow: 0;
+  }
+
+  .euiAccordion__optionalAction {
+    flex-grow: 1;
+    flex-shrink: 1;
+  }
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/StepSeparator.tsx]: We have done this in places in Kibana to get around constraints of what their API provides vs. what we need components to do. I agree it isn't ideal.

I was able to remove part of this override by doing an inline `style` using `buttonProps` that `EuiAccordion` provides.

You can see this here: 102775920f32f8d01b4b2d5165f51780666b7aab.

I'm unsure if there's another way around this in order to utilize `EuiAccordion` and make it match the design.

**Comment by justinkambic** [comment]
> Just a minor style issue when viewport width is smaller than 850px (the exnpanded step having wider width than the header)

Great call-out @kyungeunni! I've fixed this and a text overflow wrap issue I discovered while researching the problem, check out c048e8c8af21b59e1b89ffabe5a7f235ca71e7b2.

**Comment by justinkambic** [comment]
> Once you click to replay steps, IMO we should immediately disable the button because it takes a little bit of time for the browser to open, and that could cause the user to click there multiple times.

I am +1 for adding this constraint. I'll need to investigate how we can make it happen without unwanted side effects.

> I could consistently reproduce a case in which my last step was not reproduced. Please see the example below where I had one last click on a piece of text that was never reported back.

@lucasfcosta I think what you're showing me here is actually working as designed. The items in this list should correspond to the `Step 1`, `Step 2`, so there ought to only be two items displayed. If any code within a given step fails, that entire step's code is displayed today. We have some work to do in making this part of the experience a little nicer and more specific, but we are bound by some limitations related to Playwright today. In the future, if we make our usage of PW more sophisticated, and as they add additional features, we will likely be able to make this a richer part of the flow.

> Can we hide the trailing lines after the last step on each block? Please see the trailing lines after the dots on the left in the screenshots below.

Certainly can. I'm not sure if it's part of the original design, but shouldn't be too difficult. Today we omit the dots on Assertion elements. We'd probably want design input before we change anything on that.

**Review by lucasfcosta — COMMENTED** [review]
Hey Justin, this is looking excellent.

I've just tested E2E and it LGTM, although I wanted to clarify the issue I reported about the last action I've performed not appearing in the script replay (which persists, but I'm happy to solve it separately).

Essentially, in the screenshot below you can see that `step 2` has two actions:
* Clicking a link
* Clicking a piece of text

Screenshot 2022-02-11 at 17 51 50

On the flyout in the right, however, only the first action appears under the actions executed in `step 2`:

Screenshot 2022-02-11 at 17 51 50

I thought both actions should appear under `step 2`. Please let me know if I misunderstood.

------

Also, the expansion animation seems a bit weird. It looks like the expansion is faster than the element positioned above the expanding item.

weird_animation

Sorry for not having caught this earlier.

File: src/components/StepSeparator.tsx [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+      initialIsOpen
+    >
+      {step.map((s, actionIndex) => (
+        <ActionElement
+          key={`action-${actionIndex}-for-step-${index}`}
+          step={s}
+          actionIndex={actionIndex}
+          stepIndex={index}
+          testStatus={testStatus}
+          isLast={actionIndex === step.length - 1}
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/components/StepSeparator.tsx:89]: Thanks - at first I worried it was kinda hacky, but in hindsight it works rather well.

**Comment by justinkambic** [comment]
> Essentially, in the screenshot below you can see that step 2 has two actions

@lucasfcosta your confusion is warranted, this is definitely something we can clean up. We should show all of the action names.

Right now the step derives its "title" from the first action in the list, so that's what's showing up here. I think we should put this in a follow-up ticket and do it in a subsequent change so we can keep it a bit more compartmentalized. LMK if you agree or would rather fix this here.

**Comment by justinkambic** [comment]
> Also, the expansion animation seems a bit weird. It looks like the expansion is faster than the element positioned above the expanding item.

I think this is kinda strange-looking as well. I don't think there's any API for changing the animation type/speed. We could probably override it with some custom CSS. Might be another good follow-up candidate, as I don't think it's ruining the functionality of the feature but could be made to look nicer. We could also investigate with the EUI team.

I think the EUI version we're using is not compatible with the version of React we're using, so it's also possible there could be some wonky dependency stuff happening that we could resolve by bumping EUI.

**Comment by lucasfcosta** [comment]
@justinkambic, thanks for taking the time to explain both items I highlighted.

Considering what you've said, I'm happy to create separate issues for those and then press the merge button 😄

**Review by lucasfcosta — APPROVED** [review]
